File: code/examples/python/10_production_agent.py
"""
生产级 Agent 模板
=================
功能清单：
  - FastAPI 应用（POST /chat、GET /health）
  - 异步 Agent（asyncio）
  - 并发会话管理（字典存储多用户）
  - 结构化日志（trace_id、user_id、token_count）
  - 请求追踪（x-trace-id header）
  - 配置管理（pydantic BaseSettings + .env）
  - Graceful shutdown
  - 错误处理中间件

运行方式：
  pip install fastapi uvicorn anthropic pydantic-settings python-dotenv
  uvicorn 10_production_agent:app --reload --port 8000

测试：
  curl -X POST http://localhost:8000/chat \
       -H "Content-Type: application/json" \
       -H "x-trace-id: my-trace-001" \
       -d '{"user_id": "alice", "message": "你好，请介绍一下你自己"}'

  curl http://localhost:8000/health
"""

from __future__ import annotations

import asyncio
import json
import logging
import signal
import sys
import time
import uuid
from contextlib import asynccontextmanager
from datetime import datetime
from typing import Any

import anthropic
import uvicorn
from fastapi import FastAPI, HTTPException, Request, Response
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from pydantic_settings import BaseSettings, SettingsConfigDict

# ---------------------------------------------------------------------------
# 1. 配置管理（pydantic BaseSettings）
# ---------------------------------------------------------------------------

class Settings(BaseSettings):
    """从环境变量 / .env 文件读取配置。"""

    model_config = SettingsConfigDict(
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore",
    )

    # Anthropic
    anthropic_api_key: str = ""
    model: str = "claude-opus-4-5"
    max_tokens: int = 1024
    system_prompt: str = "你是一个有帮助的 AI 助手，请用简洁清晰的中文回答问题。"

    # 服务
    host: str = "0.0.0.0"
    port: int = 8000
    log_level: str = "INFO"
    max_history_turns: int = 20          # 每个会话保留的最大轮次
    session_ttl_seconds: int = 3600      # 会话超时时间（秒）


settings = Settings()


# ---------------------------------------------------------------------------
# 2. 结构化日志
# ---------------------------------------------------------------------------

class StructuredLogger:
    """输出 JSON 格式日志，包含 trace_id、user_id、token_count 等字段。"""

    def __init__(self, name: str) -> None:
        self._logger = logging.getLogger(name)
        if not self._logger.handlers:
            handler = logging.StreamHandler(sys.stdout)
            handler.setFormatter(logging.Formatter("%(message)s"))
            self._logger.addHandler(handler)
        self._logger.setLevel(getattr(logging, settings.log_level.upper(), logging.INFO))

    def _emit(self, level: str, event: str, **kwargs: Any) -> None:
        record = {
            "ts": datetime.utcnow().isoformat() + "Z",
            "level": level,
            "event": event,
            **kwargs,
        }
        self._logger.log(
            getattr(logging, level),
            json.dumps(record, ensure_ascii=False),
        )

    def info(self, event: str, **kwargs: Any) -> None:
        self._emit("INFO", event, **kwargs)

    def warning(self, event: str, **kwargs: Any) -> None:
        self._emit("WARNING", event, **kwargs)

    def error(self, event: str, **kwargs: Any) -> None:
        self._emit("ERROR", event, **kwargs)


logger = StructuredLogger("production_agent")


# ---------------------------------------------------------------------------
# 3. 会话管理
# ---------------------------------------------------------------------------

class Session:
    """单个用户的对话会话。"""

    def __init__(self, user_id: str) -> None:
        self.user_id = user_id
        self.history: list[dict[str, str]] = []
        self.created_at = time.time()
        self.last_active = time.time()
        self.total_input_tokens = 0
        self.total_output_tokens = 0

    def add_turn(self, role: str, content: str) -> None:
        self.history.append({"role": role, "content": content})
        self.last_active = time.time()
        # 超出限制时裁剪最早的轮次（保留系统完整性：成对裁剪）
        while len(self.history) > settings.max_history_turns * 2:
            self.history.pop(0)
            self.history.pop(0)

    def is_expired(self) -> bool:
        return (time.time() - self.last_active) > settings.session_ttl_seconds

    def stats(self) -> dict[str, Any]:
        return {
            "user_id": self.user_id,
            "turns": len(self.history) // 2,
            "total_input_tokens": self.total_input_tokens,
            "total_output_tokens": self.total_output_tokens,
            "created_at": datetime.utcfromtimestamp(self.created_at).isoformat() + "Z",
            "last_active": datetime.utcfromtimestamp(self.last_active).isoformat() + "Z",
        }


class SessionStore:
    """线程安全（asyncio）的多用户会话存储。"""

    def __init__(self) -> None:
        self._sessions: dict[str, Session] = {}
        self._lock = asyncio.Lock()

    async def get_or_create(self, user_id: str) -> Session:
        async with self._lock:
            session = self._sessions.get(user_id)
            if session is None or session.is_expired():
                session = Session(user_id)
                self._sessions[user_id] = session
                logger.info("session_created", user_id=user_id)
            return session

    async def delete(self, user_id: str) -> None:
        async with self._lock:
            self._sessions.pop(user_id, None)

    async def purge_expired(self) -> int:
        async with self._lock:
            expired = [uid for uid, s in self._sessions.items() if s.is_expired()]
            for uid in expired:
                del self._sessions[uid]
            return len(expired)

    def count(self) -> int:
        return len(self._sessions)


session_store = SessionStore()


# ---------------------------------------------------------------------------
# 4. 异步 Agent
# ---------------------------------------------------------------------------

class ProductionAgent:
    """异步 Agent，使用 Anthropic Messages API。"""

    def __init__(self) -> None:
        self._client = anthropic.AsyncAnthropic(
            api_key=settings.anthropic_api_key or None  # None → 读取环境变量
        )

    async def chat(
        self,
        session: Session,
        user_message: str,
        trace_id: str,
    ) -> str:
        session.add_turn("user", user_message)

        t0 = time.perf_counter()
        response = await self._client.messages.create(
            model=settings.model,
            max_tokens=settings.max_tokens,
            system=settings.system_prompt,
            messages=session.history,
        )
        elapsed_ms = int((time.perf_counter() - t0) * 1000)

        assistant_text = response.content[0].text
        session.add_turn("assistant", assistant_text)

        # 累计 token 计数
        input_tokens = response.usage.input_tokens
        output_tokens = response.usage.output_tokens
        session.total_input_tokens += input_tokens
        session.total_output_tokens += output_tokens

        logger.info(
            "agent_response",
            trace_id=trace_id,
            user_id=session.user_id,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            elapsed_ms=elapsed_ms,
            model=settings.model,
        )

        return assistant_text


agent = ProductionAgent()


# ---------------------------------------------------------------------------
# 5. 后台清理任务
# ---------------------------------------------------------------------------

async def _cleanup_loop() -> None:
    """每 5 分钟清理一次过期会话。"""
    while True:
        await asyncio.sleep(300)
        removed = await session_store.purge_expired()
        if removed:
            logger.info("sessions_purged", count=removed, active=session_store.count())


# ---------------------------------------------------------------------------
# 6. FastAPI 生命周期（含 Graceful Shutdown）
# ---------------------------------------------------------------------------

_background_tasks: list[asyncio.Task] = []


@asynccontextmanager
async def lifespan(app: FastAPI):
    """应用启动 / 关闭钩子。"""
    logger.info("app_starting", model=settings.model, port=settings.port)

    # 启动后台任务
    task = asyncio.create_task(_cleanup_loop())
    _background_tasks.append(task)

    # 注册系统信号以支持 graceful shutdown
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, lambda: asyncio.create_task(_shutdown()))

    yield  # 应用运行期间

    # 关闭时取消后台任务
    for t in _background_tasks:
        t.cancel()
        try:
            await t
        except asyncio.CancelledError:
            pass
    logger.info("app_stopped")


async def _shutdown() -> None:
    logger.info("graceful_shutdown_initiated")
    # 给进行中的请求最多 5 秒完成
    await asyncio.sleep(5)
    sys.exit(0)


# ---------------------------------------------------------------------------
# 7. FastAPI 应用与中间件
# ---------------------------------------------------------------------------

app = FastAPI(
    title="Production Agent API",
    version="1.0.0",
    lifespan=lifespan,
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)


# 错误处理中间件
@app.middleware("http")
async def error_and_tracing_middleware(request: Request, call_next):
    # 生成或透传 trace_id
    trace_id = request.headers.get("x-trace-id") or str(uuid.uuid4())
    request.state.trace_id = trace_id

    t0 = time.perf_counter()
    try:
        response: Response = await call_next(request)
        elapsed_ms = int((time.perf_counter() - t0) * 1000)
        logger.info(
            "http_request",
            trace_id=trace_id,
            method=request.method,
            path=request.url.path,
            status=response.status_code,
            elapsed_ms=elapsed_ms,
        )
        response.headers["x-trace-id"] = trace_id
        return response
    except Exception as exc:
        elapsed_ms = int((time.perf_counter() - t0) * 1000)
        logger.error(
            "unhandled_exception",
            trace_id=trace_id,
            path=request.url.path,
            error=str(exc),
            elapsed_ms=elapsed_ms,
        )
        return JSONResponse(
            status_code=500,
            content={"error": "internal_server_error", "trace_id": trace_id},
            headers={"x-trace-id": trace_id},
        )


# ---------------------------------------------------------------------------
# 8. 请求 / 响应模型
# ---------------------------------------------------------------------------

class ChatRequest(BaseModel):
    user_id: str = Field(..., description="用户唯一标识")
    message: str = Field(..., min_length=1, description="用户输入的消息")
    reset: bool = Field(False, description="是否重置当前会话历史")


class ChatResponse(BaseModel):
    user_id: str
    reply: str
    trace_id: str
    session_stats: dict[str, Any]


class HealthResponse(BaseModel):
    status: str
    active_sessions: int
    model: str
    timestamp: str


# ---------------------------------------------------------------------------
# 9. 路由
# ---------------------------------------------------------------------------

@app.post("/chat", response_model=ChatResponse)
async def chat_endpoint(body: ChatRequest, request: Request) -> ChatResponse:
    trace_id: str = getattr(request.state, "trace_id", str(uuid.uuid4()))

    logger.info(
        "chat_request",
        trace_id=trace_id,
        user_id=body.user_id,
        message_len=len(body.message),
    )

    if body.reset:
        await session_store.delete(body.user_id)
        logger.info("session_reset", trace_id=trace_id, user_id=body.user_id)

    session = await session_store.get_or_create(body.user_id)

    try:
        reply = await agent.chat(session, body.message, trace_id)
    except anthropic.AuthenticationError:
        logger.error("auth_error", trace_id=trace_id, user_id=body.user_id)
        raise HTTPException(status_code=401, detail="Anthropic API key 无效或未配置")
    except anthropic.RateLimitError:
        logger.warning("rate_limit", trace_id=trace_id, user_id=body.user_id)
        raise HTTPException(status_code=429, detail="请求频率超限，请稍后重试")
    except anthropic.APIError as exc:
        logger.error("api_error", trace_id=trace_id, user_id=body.user_id, error=str(exc))
        raise HTTPException(status_code=502, detail=f"上游 API 错误：{exc}")

    return ChatResponse(
        user_id=body.user_id,
        reply=reply,
        trace_id=trace_id,
        session_stats=session.stats(),
    )


@app.get("/health", response_model=HealthResponse)
async def health_endpoint() -> HealthResponse:
    return HealthResponse(
        status="ok",
        active_sessions=session_store.count(),
        model=settings.model,
        timestamp=datetime.utcnow().isoformat() + "Z",
    )


@app.delete("/session/{user_id}", status_code=204)
async def delete_session(user_id: str, request: Request) -> None:
    """手动删除指定用户的会话（方便测试）。"""
    trace_id: str = getattr(request.state, "trace_id", str(uuid.uuid4()))
    await session_store.delete(user_id)
    logger.info("session_deleted_manual", trace_id=trace_id, user_id=user_id)


# ---------------------------------------------------------------------------
# 10. 直接运行入口
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    uvicorn.run(
        "10_production_agent:app",
        host=settings.host,
        port=settings.port,
        log_level=settings.log_level.lower(),
        # 生产环境改用 workers=4，关闭 reload
        reload=True,
    )

File: code/examples/node/10_production_agent.js
/**
 * 10_production_agent.js
 * =======================
 * 生产级 AI 聊天服务 —— 把 AI 能力封装成可部署的 HTTP API
 *
 * 涵盖内容：
 *   1. Express 应用框架（ESM 写法）
 *   2. POST /chat（带用户ID、会话ID，支持多轮对话）
 *   3. GET  /health（健康检查，返回 {status, uptime, sessions}）
 *   4. 会话管理（Map<sessionId, messages>，自动清理过期会话）
 *   5. 限流中间件（内存实现，防止滥用）
 *   6. 结构化日志（JSON 格式，含 traceId）
 *   7. 错误处理中间件
 *
 * 运行前：
 *   npm install express @anthropic-ai/sdk
 *   export ANTHROPIC_API_KEY=sk-ant-...
 *   node 10_production_agent.js
 *
 * 测试：
 *   curl -X POST http://localhost:3000/chat \
 *     -H "Content-Type: application/json" \
 *     -d '{"userId":"u1","sessionId":"s1","message":"你好"}'
 *
 *   curl http://localhost:3000/health
 */

import express from 'express';
import Anthropic from '@anthropic-ai/sdk';
import { randomUUID } from 'crypto';

const app = express();
const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });

// ─────────────────────────────────────────────
// 1. 结构化日志（含 traceId）
// ─────────────────────────────────────────────

/**
 * 结构化日志函数
 *
 * 什么是 traceId？
 *   每个请求进来时生成一个唯一 ID（traceId），整个请求链路
 *   的所有日志都带上这个 ID，出问题时可以"顺藤摸瓜"找到完整链路。
 *   类比：快递单号 —— 通过单号能追查整个物流过程。
 *
 * 为什么用 JSON 格式？
 *   普通文本日志人眼看得懂，但 ELK/Datadog 等工具无法自动解析。
 *   JSON 日志每个字段都有名称，工具能自动索引、聚合、告警。
 *
 * @param {'info'|'warn'|'error'} level
 * @param {string} event
 * @param {object} data
 */
function log(level, event, data = {}) {
  const entry = {
    ts: new Date().toISOString(),
    level,
    event,
    ...data,
  };
  console.log(JSON.stringify(entry));
}

// ─────────────────────────────────────────────
// 2. 会话管理
// ─────────────────────────────────────────────

/**
 * 会话存储：Map<sessionId, SessionData>
 *
 * 什么是"会话"？
 *   就像微信聊天窗口，每个窗口有独立的聊天记录。
 *   这里用内存存储，生产环境应换成 Redis。
 *
 * SessionData 结构：
 *   messages   : 消息历史数组（[{role, content}]）
 *   userId     : 用户 ID（可用于权限校验）
 *   createdAt  : 创建时间（毫秒时间戳）
 *   lastActive : 最后活跃时间（毫秒时间戳）
 */
const sessions = new Map();

const SESSION_TTL_MS = 30 * 60 * 1000;   // 会话超时：30 分钟
const MAX_MESSAGES_PER_SESSION = 50;       // 每个会话最多保存 50 条消息

/**
 * 获取或创建会话（懒加载）
 */
function getOrCreateSession(sessionId, userId) {
  if (!sessions.has(sessionId)) {
    sessions.set(sessionId, {
      messages: [],
      userId,
      createdAt: Date.now(),
      lastActive: Date.now(),
    });
    log('info', 'session_created', { sessionId, userId });
  }
  const session = sessions.get(sessionId);
  session.lastActive = Date.now(); // 每次访问更新活跃时间
  return session;
}

/**
 * 定期清理过期会话（每 5 分钟运行一次）
 * 类比：餐厅服务员定期清理长时间无人的桌子，腾出空间
 */
function cleanupSessions() {
  const now = Date.now();
  let cleaned = 0;

  for (const [sessionId, session] of sessions.entries()) {
    if (now - session.lastActive > SESSION_TTL_MS) {
      sessions.delete(sessionId);
      cleaned++;
    }
  }

  if (cleaned > 0) {
    log('info', 'sessions_cleaned', { cleaned, remaining: sessions.size });
  }
}

setInterval(cleanupSessions, 5 * 60 * 1000);

// ─────────────────────────────────────────────
// 3. 限流中间件（内存实现）
// ─────────────────────────────────────────────

/**
 * 简单的滑动窗口限流
 *
 * 原理：记录每个 userId 在最近 windowMs 内的请求次数，
 *       超过 maxRequests 就返回 429。
 *
 * 类比：银行取号机每小时最多给你取 3 个号，超了就让你等。
 *
 * 生产环境推荐：npm install express-rate-limit，配合 Redis 存储做分布式限流
 *
 * @param {number} windowMs    - 统计窗口（ms，默认60秒）
 * @param {number} maxRequests - 窗口内最大请求数（默认20）
 */
function rateLimitMiddleware(windowMs = 60000, maxRequests = 20) {
  const store = new Map(); // Map<userId, number[]>（请求时间戳数组）

  return (req, res, next) => {
    const userId = req.body?.userId || req.ip || 'anonymous';
    const now = Date.now();
    const windowStart = now - windowMs;

    // 只保留窗口内的时间戳
    const timestamps = (store.get(userId) || []).filter(ts => ts > windowStart);

    if (timestamps.length >= maxRequests) {
      const resetAt = new Date(timestamps[0] + windowMs).toISOString();
      log('warn', 'rate_limit_exceeded', {
        userId,
        requestCount: timestamps.length,
        resetAt,
      });
      return res.status(429).json({
        error: '请求太频繁，请稍后再试',
        retryAfterSeconds: Math.ceil((timestamps[0] + windowMs - now) / 1000),
        resetAt,
      });
    }

    timestamps.push(now);
    store.set(userId, timestamps);
    next();
  };
}

// ─────────────────────────────────────────────
// 4. 中间件配置
// ─────────────────────────────────────────────

app.use(express.json({ limit: '16kb' })); // 限制请求体大小

// 请求日志中间件：为每个请求附加 traceId
app.use((req, res, next) => {
  req.traceId = randomUUID(); // 每次请求生成唯一 ID
  req.startTime = Date.now();

  log('info', 'request_received', {
    traceId: req.traceId,
    method: req.method,
    path: req.path,
  });

  // 响应结束后记录耗时（利用 Node.js 事件钩子）
  res.on('finish', () => {
    log('info', 'request_completed', {
      traceId: req.traceId,
      statusCode: res.statusCode,
      durationMs: Date.now() - req.startTime,
    });
  });

  next();
});

// ─────────────────────────────────────────────
// 5. 路由：POST /chat
// ─────────────────────────────────────────────

/**
 * POST /chat
 *
 * 请求体（JSON）：
 *   { userId: string, sessionId: string, message: string }
 *
 * 成功响应（200）：
 *   { reply: string, sessionId: string, messageCount: number, usage: {...} }
 *
 * 错误响应：
 *   400 - 参数缺失或无效
 *   429 - 请求频率超限
 *   503 - AI 服务不可用
 */
app.post('/chat',
  rateLimitMiddleware(60000, 20), // 每分钟最多 20 条消息
  async (req, res, next) => {
    const { userId, sessionId, message } = req.body;

    // 参数校验
    if (!userId || !sessionId || !message) {
      return res.status(400).json({
        error: '缺少必要参数',
        required: ['userId', 'sessionId', 'message'],
        traceId: req.traceId,
      });
    }
    if (typeof message !== 'string' || message.trim().length === 0) {
      return res.status(400).json({ error: 'message 不能为空', traceId: req.traceId });
    }
    if (message.length > 10000) {
      return res.status(400).json({ error: 'message 不能超过 10000 字符', traceId: req.traceId });
    }

    try {
      const session = getOrCreateSession(sessionId, userId);

      // 添加用户消息
      session.messages.push({ role: 'user', content: message.trim() });

      // 消息超出上限时截断（保留最近的，防止内存无限增长）
      if (session.messages.length > MAX_MESSAGES_PER_SESSION) {
        const removed = session.messages.length - MAX_MESSAGES_PER_SESSION;
        session.messages = session.messages.slice(-MAX_MESSAGES_PER_SESSION);
        log('warn', 'session_truncated', { sessionId, removed, traceId: req.traceId });
      }

      log('info', 'calling_anthropic', {
        traceId: req.traceId,
        userId,
        sessionId,
        messageCount: session.messages.length,
      });

      // 调用 Anthropic API
      const response = await client.messages.create({
        model: 'claude-opus-4-5',
        max_tokens: 1024,
        system: '你是一个友好的 AI 助手，用中文回答用户问题。',
        messages: session.messages,
      });

      const reply = response.content[0].text;

      // 把 AI 回复也加入历史（维持多轮对话上下文）
      session.messages.push({ role: 'assistant', content: reply });

      log('info', 'chat_success', {
        traceId: req.traceId,
        userId,
        sessionId,
        inputTokens: response.usage.input_tokens,
        outputTokens: response.usage.output_tokens,
      });

      res.json({
        reply,
        sessionId,
        messageCount: session.messages.length,
        usage: {
          inputTokens: response.usage.input_tokens,
          outputTokens: response.usage.output_tokens,
        },
        traceId: req.traceId,
      });
    } catch (error) {
      next(error); // 传递给全局错误处理中间件
    }
  }
);

// ─────────────────────────────────────────────
// 6. 路由：GET /health
// ─────────────────────────────────────────────

/**
 * GET /health
 *
 * 健康检查端点，供负载均衡器和监控系统调用。
 * 类比：体检 —— 快速检查服务是否"活着"且状态正常。
 *
 * 负载均衡器（如 Nginx、AWS ALB）会定期请求这个接口，
 * 返回非 2xx 或超时时，会把该实例从流量池中摘除。
 *
 * 响应 200：
 *   { status: "ok", uptime: 123.45, sessions: 5 }
 */
app.get('/health', (req, res) => {
  const memUsage = process.memoryUsage();
  res.json({
    status: 'ok',
    uptime: process.uptime(),           // 进程运行时间（秒）
    sessions: sessions.size,            // 当前活跃会话数
    memory: {
      heapUsed: Math.round(memUsage.heapUsed / 1024 / 1024) + 'MB',
      rss: Math.round(memUsage.rss / 1024 / 1024) + 'MB',
    },
    version: '1.0.0',
    timestamp: new Date().toISOString(),
  });
});

// ─────────────────────────────────────────────
// 7. 错误处理中间件
// ─────────────────────────────────────────────

/**
 * 全局错误处理中间件
 *
 * Express 规定：错误中间件必须有 4 个参数（err, req, res, next），
 * 否则 Express 不会把它识别为错误处理器。
 * 位置必须放在所有路由之后。
 */
// eslint-disable-next-line no-unused-vars
app.use((err, req, res, next) => {
  const traceId = req.traceId || 'unknown';

  // Anthropic API 错误细分处理
  if (err instanceof Anthropic.APIError) {
    log('error', 'anthropic_api_error', {
      traceId,
      status: err.status,
      message: err.message,
    });

    if (err.status === 429) {
      return res.status(503).json({
        error: 'AI 服务暂时繁忙，请稍后再试',
        retryAfterSeconds: 5,
        traceId,
      });
    }
    if (err.status >= 500) {
      return res.status(503).json({ error: 'AI 服务暂时不可用', traceId });
    }
  }

  // 其他错误
  log('error', 'internal_error', {
    traceId,
    message: err.message,
    // 生产环境不打印 stack 到响应，避免泄露内部实现
    stack: process.env.NODE_ENV === 'development' ? err.stack : undefined,
  });

  res.status(500).json({
    error: '服务器内部错误',
    traceId,
    detail: process.env.NODE_ENV === 'development' ? err.message : undefined,
  });
});

// ─────────────────────────────────────────────
// 8. 启动服务器
// ─────────────────────────────────────────────

const PORT = process.env.PORT || 3000;

app.listen(PORT, () => {
  log('info', 'server_started', {
    port: PORT,
    env: process.env.NODE_ENV || 'development',
    apiKeySet: !!process.env.ANTHROPIC_API_KEY,
  });

  console.log(`
╔═══════════════════════════════════════════╗
║  10_production_agent.js  生产级 AI 服务   ║
╠═══════════════════════════════════════════╣
║  服务已启动：http://localhost:${PORT}         ║
║                                           ║
║  端点：                                   ║
║    POST /chat    多轮对话                  ║
║    GET  /health  健康检查                  ║
║                                           ║
║  测试 chat：                              ║
║  curl -X POST http://localhost:${PORT}/chat \\║
║    -H "Content-Type: application/json" \\ ║
║    -d '{"userId":"u1","sessionId":"s1",   ║
║         "message":"你好"}'                ║
║                                           ║
║  测试 health：                            ║
║  curl http://localhost:${PORT}/health        ║
╚═══════════════════════════════════════════╝
  `);
});

// 优雅关闭：SIGTERM 时等待当前请求完成再退出（Docker/K8s 标准做法）
process.on('SIGTERM', () => {
  log('info', 'server_shutting_down', { activeSessions: sessions.size });
  process.exit(0);
});
